src/sql.rs

```diff
@@ -53,3 +53,39 @@ macro_rules! pg_enum {
 }
 
 pub(crate) use pg_enum;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::db::sql_types::semver::Triple;
+    use crate::schema::sql_types::SemverTriple;
+    use crate::test_util::pg_connection;
+    use diesel::prelude::*;
+    use diesel::select;
+
+    sql_function!(fn to_semver_no_prerelease(x: Text) -> Nullable<SemverTriple>);
+
+    #[test]
+    fn to_semver_no_prerelease_works() {
+        let mut conn = pg_connection();
+
+        #[track_caller]
+        fn test(conn: &mut PgConnection, text: &str, expected: Option<(u64, u64, u64)>) {
+            let query = select(to_semver_no_prerelease(text));
+            let result = query
+                .get_result::<Option<Triple>>(conn)
+                .unwrap()
+                .map(|triple| (triple.major, triple.minor, triple.teeny));
+
+            assert_eq!(result, expected);
+        }
+
+        test(&mut conn, "0.0.0", Some((0, 0, 0)));
+        test(&mut conn, "1.2.4", Some((1, 2, 4)));
+        test(&mut conn, "1.2.4+metadata", Some((1, 2, 4)));
+        test(&mut conn, "1.2.4-beta.3", None);
+        // see https://github.com/rust-lang/crates.io/issues/3882
+        test(&mut conn, "0.4.45+curl-7.78.0", Some((0, 4, 45)));
+        test(&mut conn, "0.1.4-preview+4.3.2", None);
```

Comment:
What about `1.0.0-beta+exp.sha.5114f85`? It seems we can not handle it right now. But it is a valid version in [regex101.com/r/vEYRnS/1](https://regex101.com/r/vEYRnS/1).

Reply:
I also tested it with semver:

```rust
#!/usr/bin/env -S cargo +nightly -Zscript

//! ```cargo
//! [dependencies]
//! semver = "1.0"
//! ```


use semver::Version;

fn main() {
    match Version::parse("1.0.0-beta+exp.sha.5114f85") {
        Ok(v) => println!("Valid version: {}", v),
        Err(e) => println!("Invalid version: {}", e),
    }
}
```

```console
cargo +nightly -Zscript version.rs
warning: `package.edition` is unspecifiead, defaulting to `2021`
   Compiling version v0.0.0 (/Volumes/t7/code/cargo-scripts)
    Finished dev [unoptimized + debuginfo] target(s) in 0.39s
     Running `/Users/hi-rustin/.cargo/target/bd/d06c14836752f5/debug/version`
Valid version: 1.0.0-beta+exp.sha.5114f85
```